File: 16.observer/observer.py
# Sujeto
class Subject:
    def __init__(self):
        self.observers = []

    def subscribe(self, observer):
        self.observers.append(observer)

    def unsubscribe(self, observer):
        self.observers.remove(observer)

    def notify_observers(self, offer):
        for observer in self.observers:
            observer.update(offer)

# Observador
class User:
    def __init__(self, name):
        self.name = name

    def update(self, offer):
        print(f"{self.name} recibió la notificación: Nueva oferta - {offer}")

# Uso
store = Subject()

user1 = User("Alice")
user2 = User("Bob")

store.subscribe(user1)
store.subscribe(user2)

store.notify_observers("50% de descuento en laptops")

File: 16.observer/observer.java
// Observador
interface Observer {
    void update(String offer);
}

// Sujeto
class Subject {
    private List<Observer> observers = new ArrayList<>();

    public void subscribe(Observer observer) {
        observers.add(observer);
    }

    public void unsubscribe(Observer observer) {
        observers.remove(observer);
    }

    public void notifyObservers(String offer) {
        for (Observer observer : observers) {
            observer.update(offer);
        }
    }
}

// Observador concreto
class User implements Observer {
    private String name;

    public User(String name) {
        this.name = name;
    }

    @Override
    public void update(String offer) {
        System.out.println(name + " recibió la notificación: Nueva oferta - " + offer);
    }
}

// Uso
public class observer {
    public static void main(String[] args) {
        Subject store = new Subject();

        Observer user1 = new User("Alice");
        Observer user2 = new User("Bob");

        store.subscribe(user1);
        store.subscribe(user2);

        store.notifyObservers("50% de descuento en laptops");
    }
}